File: apps/api/app/schemas/ask.py
"""Ask (grounded RAG) request/response schemas."""

from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class AskRequest(BaseModel):
    question: str = Field(min_length=1, max_length=2000)
    language: str | None = None
    path_prefix: str | None = None
    support_level: str | None = None
    apply_rerank: bool = True
    expand: bool = True


class AskCitation(BaseModel):
    path: str
    start_line: int
    end_line: int
    chunk_id: UUID | None = None
    valid: bool = True
    reason: str = "ok"
    raw: str | None = None


class AskEvidenceItem(BaseModel):
    chunk_id: UUID
    path: str
    start_line: int
    end_line: int
    support_level: str
    role: str
    depth: int
    citation: str
    evidence_tier: int | None = None
    chars_stored: int | None = None
    chars_sent: int | None = None
    content_truncated: bool | None = None
    estimated_tokens_sent: int | None = None


class AskAnalysisEcho(BaseModel):
    kind: str
    rewrite_applied: bool
    retrieval_queries: list[str] = Field(default_factory=list)
    identifiers: list[str] = Field(default_factory=list)
    paths: list[str] = Field(default_factory=list)


class AskValidationEcho(BaseModel):
    ok: bool
    valid_count: int
    dropped_count: int
    errors: list[str] = Field(default_factory=list)


class AskBudgetEcho(BaseModel):
    requests_used: int
    requests_limit: int
    tokens_used: int
    tokens_limit: int
    estimated_cost_usd: float
    cost_limit_usd: float
    exhausted: bool = False
    skipped_reason: str | None = None
    remaining_requests: int = 0


class AskFileCoverage(BaseModel):
    requested_file: str | None = None
    path: str | None = None
    indexed: bool = False
    coverage_complete: bool = False
    indexed_line_range: list[int] | None = None
    retrieved_line_ranges: list[list[int]] = Field(default_factory=list)
    missing_ranges: list[list[int]] = Field(default_factory=list)
    chunk_count: int = 0
    language: str | None = None
    classification: str | None = None


class AskFileDiagnostic(BaseModel):
    path: str
    indexed: bool = False
    source_file_present: bool = False
    skip_reason: str | None = None
    classification: str | None = None
    language: str | None = None
    chunk_count: int = 0
    line_coverage: list[list[int]] | None = None
    indexed_line_range: list[int] | None = None
    coverage_complete: bool | None = None
    missing_ranges: list[list[int]] | None = None
    retrieval_reason: str | None = None
    basename_aliases: list[str] = Field(default_factory=list)
    snapshot_id: str | None = None


class AskResponse(BaseModel):
    repository_id: UUID
    snapshot_id: UUID | None
    question: str
    answer: str
    status: str
    citations: list[AskCitation] = Field(default_factory=list)
    evidence: list[AskEvidenceItem] = Field(default_factory=list)
    analysis: AskAnalysisEcho | None = None
    validation: AskValidationEcho
    context_depth: int = 0
    context_tokens_used: int = 0
    context_token_budget: int = 0
    ranked_chunk_ids: list[UUID] = Field(default_factory=list)
    model_provenance: dict[str, Any] | None = None
    cached: bool = False
    notes: list[str] = Field(default_factory=list)
    budget: AskBudgetEcho | None = None
    context_truncated: bool = False
    exact_file_mode: bool = False
    file_coverage: list[AskFileCoverage] = Field(default_factory=list)
    file_diagnostics: list[AskFileDiagnostic] = Field(default_factory=list)

File: apps/web/src/lib/ask.ts
export type AskStatus =
  | "ok"
  | "partial"
  | "no_evidence"
  | "ask_disabled"
  | "budget_exceeded"
  | "error"
  | string;

export interface AskCitation {
  path: string;
  start_line: number;
  end_line: number;
  chunk_id: string | null;
  valid: boolean;
  reason: string;
  raw: string | null;
}

export interface AskEvidenceItem {
  chunk_id: string;
  path: string;
  start_line: number;
  end_line: number;
  support_level: string;
  role: string;
  depth: number;
  citation: string;
}

export interface AskAnalysisEcho {
  kind: string;
  rewrite_applied: boolean;
  retrieval_queries: string[];
  identifiers: string[];
  paths: string[];
}

export interface AskValidationEcho {
  ok: boolean;
  valid_count: number;
  dropped_count: number;
  errors: string[];
}

export interface AskBudgetEcho {
  requests_used: number;
  requests_limit: number;
  tokens_used: number;
  tokens_limit: number;
  estimated_cost_usd: number;
  cost_limit_usd: number;
  exhausted: boolean;
  skipped_reason: string | null;
  remaining_requests: number;
}

export interface AskRequestBody {
  question: string;
  language?: string;
  path_prefix?: string;
  support_level?: string;
  apply_rerank?: boolean;
  expand?: boolean;
}

export interface AskResponse {
  repository_id: string;
  snapshot_id: string | null;
  question: string;
  answer: string;
  status: AskStatus;
  citations: AskCitation[];
  evidence: AskEvidenceItem[];
  analysis: AskAnalysisEcho | null;
  validation: AskValidationEcho;
  context_depth: number;
  context_tokens_used: number;
  context_token_budget: number;
  ranked_chunk_ids: string[];
  model_provenance: Record<string, unknown> | null;
  cached: boolean;
  notes: string[];
  budget: AskBudgetEcho | null;
}

export function askCitationLabel(cite: {
  path: string;
  start_line: number;
  end_line: number;
}): string {
  return `${cite.path}:${cite.start_line}-${cite.end_line}`;
}

export function askStatusLabel(status: string): string {
  switch (status) {
    case "ok":
      return "OK";
    case "partial":
      return "Partial (some citations dropped)";
    case "no_evidence":
      return "No evidence";
    case "ask_disabled":
      return "Ask disabled";
    case "budget_exceeded":
      return "Repo Ask budget exhausted";
    case "error":
      return "Error";
    default:
      return status;
  }
}
